src/samplers/support/hmc_core.jl

```diff
@@ -36,13 +46,15 @@ leapfrog(_θ::Union{Vector,SubArray}, p::Vector{Float64}, τ::Int, ϵ::Float64,
 
     vi[spl] = θ
     grad = gradient(vi, model, spl)
-    verifygrad(grad) || (vi[spl] = θ_old; setlogp!(vi, old_logp); θ = θ_old; p = p_old; break)
+    verifygrad(grad) || (vi[spl] = θ_old; setbothlogp!(vi, old_logp); θ = θ_old; p = p_old; break)
 
     p -= ϵ * grad / 2
 
     τ_valid += 1
   end
 
+  θ[indep_rvs] = θ_indep_rvs
```

Comment:
The way we use `get_indep_rvs` still do the computation to those independent random variables here... Do you think we can use it outside this function and avoid doing the computation?